File: src/main/kotlin/org/core/scheduleflow/domain/partner/service/PartnerService.java
package org.core.scheduleflow.domain.partner.service;

import org.core.scheduleflow.domain.partner.dto.PartnerRequestDto;
import org.core.scheduleflow.domain.partner.dto.PartnerResponseDto;
import org.core.scheduleflow.domain.partner.entity.Partner;
import org.core.scheduleflow.domain.partner.repository.PartnerRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class PartnerService {

    private final PartnerRepository partnerRepository;

    public PartnerService(PartnerRepository partnerRepository) {
        this.partnerRepository = partnerRepository;
    }

    /*==========================================================Partner READ====================================================================*/

    public List<PartnerResponseDto> findAll(){
        List<Partner> partners = partnerRepository.findAll();

        return partners.stream()
                .map(PartnerResponseDto::fromEntity)
                .collect(Collectors.toList());
    }

    public Optional<PartnerResponseDto> selectPartnerById(Long id){
        Optional<Partner> partner = partnerRepository.findById(id);


        return partner.map(PartnerResponseDto::fromEntity);
    }

    public List<PartnerResponseDto> selectPartnerByNameContains(String name){
        List<Partner> partners = partnerRepository.selectPartnerByNameContains(name);

        return partners.stream()
                .map(PartnerResponseDto::fromEntity)
                .collect(Collectors.toList());
    }

    /*===========================================================Partner CREATE================================================================*/

    @Transactional
    public PartnerResponseDto createPartner(PartnerRequestDto requestDto){

        /* 유효성 검증 시작 */
        Objects.requireNonNull(requestDto, "요청 데이터(PartnerRequestDto)는 null일 수 없습니다.");
        if (requestDto.getCompanyName() == null || requestDto.getCompanyName().trim().isEmpty()) {
            throw new IllegalArgumentException("회사 이름은 필수 입력 항목입니다.");
        }
        /* 유효성 검증 끝 */

        Partner partner = requestDto.toEntity();

        Partner savedPartner = partnerRepository.save(partner);

        return PartnerResponseDto.fromEntity(savedPartner);
    }

    /*===========================================================Partner UPDATE================================================================*/
    
    @Transactional
    public PartnerResponseDto updatePartner(PartnerRequestDto requestDto){

        /* 유효성 검증 시작 */
        Objects.requireNonNull(requestDto, "요청 데이터(PartnerRequestDto)는 null일 수 없습니다.");
        if (requestDto.getCompanyName() == null || requestDto.getCompanyName().trim().isEmpty()) {
            throw new IllegalArgumentException("회사 이름은 필수 입력 항목입니다.");
        }
        /* 유효성 검증 끝 */

        Partner partner = requestDto.toEntity();

        Partner savedPartner = partnerRepository.save(partner);

        return PartnerResponseDto.fromEntity(savedPartner);
    }

    /*==========================================================Partner DELETE=================================================================*/

    @Transactional
    public void deletePartnerById(Long id){
        partnerRepository.deleteById(id);
    }

}

File: src/main/kotlin/org/core/scheduleflow/domain/partner/service/PartnerService.kt
package org.core.scheduleflow.domain.partner.service

import org.core.scheduleflow.domain.partner.dto.PartnerRequestDto
import org.core.scheduleflow.domain.partner.dto.PartnerResponseDto

import org.core.scheduleflow.domain.partner.repository.PartnerRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional


@Service
@Transactional
class PartnerService(private val partnerRepository: PartnerRepository) {
    /*==========================================================Partner READ====================================================================*/
    fun findAll(): List<PartnerResponseDto> {
        val partners = partnerRepository.findAll()

        return partners.mapNotNull { partner ->
            partner?.let { PartnerResponseDto.fromEntity(it) }
        }
    }

    fun findPartnerById(id: Long): PartnerResponseDto? {

        val partner = partnerRepository.findById(id).orElse(null) ?: return null

        return PartnerResponseDto.fromEntity(partner)
    }

    fun findPartnerByNameContains(companyName: String?): List<PartnerResponseDto> {
        val partners = partnerRepository.findByCompanyNameContains(companyName)

        // mapNotNull을 사용하여 null 요소는 제외하고 DTO로 변환합니다.
        if (partners != null) {
            return partners.mapNotNull { it?.let { partner -> PartnerResponseDto.fromEntity(partner) } }
        }

        return emptyList()
    }

    /*===========================================================Partner CREATE================================================================*/

    fun createPartner(partnerRequestDto: PartnerRequestDto): PartnerResponseDto {
        /* 유효성 검증 시작 */

        require(!(partnerRequestDto.companyName.trim { it <= ' ' }.isEmpty())) { "회사 이름은 필수 입력 항목입니다." }

        /* 유효성 검증 끝 */
        val partner = partnerRequestDto.toEntity()

        val savedPartner = partnerRepository.save(partner)

        return PartnerResponseDto.fromEntity(savedPartner)
    }

    /*===========================================================Partner UPDATE================================================================*/
    @Transactional
    fun updatePartner(partnerRequestDto: PartnerRequestDto): PartnerResponseDto {
        /* 유효성 검증 시작 */

        val partnerId = partnerRequestDto.id ?: throw IllegalArgumentException("수정 시 ID는 필수입니다.")

        partnerRepository.findById(partnerId).orElseThrow() {IllegalArgumentException("존재하지 않는 고객사 입니다")}


        /* 유효성 검증 끝 */
        val partner = partnerRequestDto.toEntity()

        val savedPartner = partnerRepository.save(partner)

        return PartnerResponseDto.fromEntity(savedPartner)
    }

    /*==========================================================Partner DELETE=================================================================*/
    @Transactional
    fun deletePartnerById(id: Long) {
        partnerRepository.deleteById(id)
    }
}
